# Segmenting and Clustering Neighborhoods in Toronto

###### Author: Adam Kassoff

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis


## Part 1:
1. Creat notebook
2. Use the Notebook to build the code to scrape the following Wikipedia page, https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M, in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown below: <img src='https://d3c33hcgiwev3.cloudfront.net/imageAssetProxy.v1/7JXaz3NNEeiMwApe4i-fLg_40e690ae0e927abda2d4bde7d94ed133_Screen-Shot-2018-06-18-at-7.17.57-PM.png?expiry=1603324800000&hmac=WmqqbK-twXsXwEaCPFbaPEoGST65EIKqAcXdBRU0nqQ' >

- The dataframe will consist of three columns: PostalCode, Borough, and Neighborhood
- Only process the cells that have an assigned borough. Ignore cells with a borough that is Not assigned.
- More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.
- If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough.
- Clean your Notebook and add Markdown cells to explain your work and any assumptions you are making.
- In the last cell of your notebook, use the .shape method to print the number of rows of your dataframe.

## Importing Data

In [2]:
import requests

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

Response 200 means that we are able to make the connection to the page

In [4]:
wiki_data = pd.read_html(wiki_url.content)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [5]:
len(wiki_data), type(wiki_data)

(3, list)

We only care about the first table, so lets drop the other tables.

In [6]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Now we need to drop the data with not assigned values for borough.

In [7]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df.reset_index(inplace = True)
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

Now let's group the data by the Postal Codes

In [8]:
df.groupby(['Postal Code']).first()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

Checking for number of records where Neighbourhood is "Not assigned"

In [9]:
df.Neighbourhood.str.count("Not assigned").sum()

0

Remove the redundant "index" column

In [10]:
df.drop(['index'], axis = 'columns', inplace = True)
df

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [11]:
df.shape

(103, 3)

### Answer to Part 1: We have 103 rows and 3 columns

________________________________

# Part 2

- Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood.
- Use the Geocoder package or the csv file to create the following dataframe:
<img src='https://d3c33hcgiwev3.cloudfront.net/imageAssetProxy.v1/HZ3jNHNOEeiMwApe4i-fLg_f44f0f10ccfaf42fcbdba9813364e173_Screen-Shot-2018-06-18-at-7.18.16-PM.png?expiry=1603324800000&hmac=pAe0lBDXVwV8osg8Q1YXCqKpMKtDWT0wt2myXP1-99k'>

Before we import geocoder we have to install geocoder

In [12]:
pip install geocoder

     |████████████████████████████████| 98 kB 6.8 MB/s eta 0:00:011
Note: you may need to restart the kernel to use updated packages.


In [13]:
import geocoder

The following code didn't work and seemed to be stuck in an infinte loop.
>#initialize your variable to None <br>
lat_lng_coords = None
>
>postal_code = 'MG5'
>
>#loop until you get the coordinates <br>
>while(lat_lng_coords is None): <br>
>  g = geocoder.google('{}, Toronto, Ontario'.format(postal_code)) <br>
>  lat_lng_coords = g.latlng <br>
>
>latitude = lat_lng_coords[0] <br>
longitude = lat_lng_coords[1]

#### Let's try using the csv file

In [14]:
df_geo = pd.read_csv("https://cocl.us/Geospatial_data")
df_geo.head

<bound method NDFrame.head of     Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]>

Let's check the data types of our columns specifically the postal codes, so we can combine the two data frames

In [15]:
df_geo.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [16]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

Let's also check that both the shapes match so no postal codes are missing

In [17]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", df_geo.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


### Here is were we join both data frames to get our desired result for part 2

In [18]:
combined_data = df.join(df_geo.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [19]:
combined_data.shape

(103, 5)

______

# Part 3

#### Explore and cluster the neighborhoods in Toronto. You can decide to work with only boroughs that contain the word Toronto and then replicate the same analysis we did to the New York City data. It is up to you.

In [20]:
from geopy.geocoders import Nominatim

address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


Let's vizualize the map of Toronto

In [22]:
!pip install folium
import folium

     |████████████████████████████████| 93 kB 3.8 MB/s  eta 0:00:01


In [23]:
# Creating the map of Toronto
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Define Foursquare API Credentials and Version

In [176]:
# The code was removed by Watson Studio for sharing.

Let's explore the data, and get the venues in 500 meters range from our first entry

In [32]:
neighbourhood_latitude = combined_data.loc[0, 'Latitude'] # neighbourhood latitude value
neighbourhood_longitude = combined_data.loc[0, 'Longitude'] # neighbourhood longitude value

neighbourhood_name = combined_data.loc[0, 'Neighbourhood'] # neighbourhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


Create the GET request URL

In [143]:
LIMIT = 100 # limit of number of venues returned by Foursquare API

radius = 500 # define radius

# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighbourhood_latitude, 
    neighbourhood_longitude, 
    radius, 
    LIMIT)
url # display URL

'https://api.foursquare.com/v2/venues/explore?&client_id=FFN1U3J0RP1BGN41N50XKRUTNJPJQ1WH1Y1SSPGTB5M3Y3F2&client_secret=QBRMF3YWOD2OO1W1ZWISMUXGWZ4WQYBABNRT5DGCBFTCFH23&v=20180605&ll=43.7532586,-79.3296565&radius=500&limit=100'

In [144]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5f8f820cc0877943bc1dac92'},
 'response': {'warning': {'text': "There aren't a lot of results near you. Try something more general, reset your filters, or expand the search area."},
  'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 2,
  'suggestedBounds': {'ne': {'lat': 43.757758604500005,
    'lng': -79.32343823984928},
   'sw': {'lat': 43.7487585955, 'lng': -79.33587476015072}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': 

In [145]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [146]:
import json
from pandas.io.json import json_normalize

venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues = nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/opt/conda/envs/Python-3.7-main/lib/python3.7/site-packages/ipykernel/__main__.py:6: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead


name         categories        lat        lng
0  Brookbanks Park               Park  43.751976 -79.332140
1    Variety Store  Food & Drink Shop  43.751974 -79.333114

Let's create a function to repeat the same process to all the neighborhoods in Toronto

In [147]:
def getNearbyVenues(names, latitudes, longitudes, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now run the above function on each neighborhood and create a new dataframe called toronto_venues.

In [148]:
toronto_venues = getNearbyVenues(names=combined_data['Neighbourhood'],
                                   latitudes=combined_data['Latitude'],
                                   longitudes=combined_data['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

Checking the size of df

In [149]:
print(toronto_venues.shape)
toronto_venues.head()

(4878, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0     Parkwoods               43.753259               -79.329656   
1     Parkwoods               43.753259               -79.329656   
2     Parkwoods               43.753259               -79.329656   
3     Parkwoods               43.753259               -79.329656   
4     Parkwoods               43.753259               -79.329656   

               Venue  Venue Latitude  Venue Longitude        Venue Category  
0    Allwyn's Bakery       43.759840       -79.324719  Caribbean Restaurant  
1    Brookbanks Park       43.751976       -79.332140                  Park  
2        Tim Hortons       43.760668       -79.326368                  Café  
3  Bruno's valu-mart       43.746143       -79.324630         Grocery Store  
4                A&W       43.760643       -79.326865  Fast Food Restaurant

Let's check how many venues were returned for each neighborhood

In [150]:
toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                             
Agincourt                                                            50   
Alderwood, Long Branch                                               24   
Bathurst Manor, Wilson Heights, Downsview North                      29   
Bayview Village                                                      14   
Bedford Park, Lawrence Manor East                                    41   
...                                                                 ...   
Willowdale, Willowdale West                                          11   
Woburn                                                                8   
Woodbine Heights                                                     26   
York Mills West                                                      20   
York Mills, Silver Hills                                              4   

                                                 Neighbourhood Longitude  \
Neighbourhood                                                              
Agincourt                                                             50   
Alderwood, Long Branch                                                24   
Bathurst Manor, Wilson Heights, Downsview North                       29   
Bayview Village                                                       14   
Bedford Park, Lawrence Manor East                                     41   
...                                                                  ...   
Willowdale, Willowdale West                                           11   
Woburn                                                                 8   
Woodbine Heights                                                      26   
York Mills West                                                       20   
York Mills, Silver Hills                                               4   

                                                 Venue  Venue Latitude  \
Neighbourhood                                                            
Agincourt                                           50              50   
Alderwood, Long Branch                              24              24   
Bathurst Manor, Wilson Heights, Downsview North     29              29   
Bayview Village                                     14              14   
Bedford Park, Lawrence Manor East                   41              41   
...                                                ...             ...   
Willowdale, Willowdale West                         11              11   
Woburn                                               8               8   
Woodbine Heights                                    26              26   
York Mills West                                     20              20   
York Mills, Silver Hills                             4               4   

                                                 Venue Longitude  \
Neighbourhood                                                      
Agincourt                                                     50   
Alderwood, Long Branch                                        24   
Bathurst Manor, Wilson Heights, Downsview North               29   
Bayview Village                                               14   
Bedford Park, Lawrence Manor East                             41   
...                                                          ...   
Willowdale, Willowdale West                                   11   
Woburn                                                         8   
Woodbine Heights                                              26   
York Mills West                                               20   
York Mills, Silver Hills                                       4   

                                                 Venue Category  
Neighbourhood                                                    
Agincourt                                                    50  
Alderwood, Long Branch                

In [151]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 334 uniques categories.


Analyze each neighbourhood

In [152]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighbourhood'] = toronto_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighbourhood  Accessories Store  Afghan Restaurant  African Restaurant  \
0     Parkwoods                  0                  0                   0   
1     Parkwoods                  0                  0                   0   
2     Parkwoods                  0                  0                   0   
3     Parkwoods                  0                  0                   0   
4     Parkwoods                  0                  0                   0   

   Airport  American Restaurant  Amphitheater  Animal Shelter  Antique Shop  \
0        0                    0             0               0             0   
1        0                    0             0               0             0   
2        0                    0             0               0             0   
3        0                    0             0               0             0   
4        0                    0             0               0             0   

   Aquarium  ...  Video Store  Vietnamese Restaurant  Warehouse Store  \
0         0  ...            0                      0                0   
1         0  ...            0                      0                0   
2         0  ...            0                      0                0   
3         0  ...            0                      0                0   
4         0  ...            0                      0                0   

   Whisky Bar  Wine Bar  Wine Shop  Wings Joint  Women's Store  Yoga Studio  \
0           0         0          0            0              0            0   
1           0         0          0            0              0            0   
2           0         0          0            0              0            0   
3           0         0          0            0              0            0   
4           0         0          0            0              0            0   

   Zoo  
0    0  
1    0  
2    0  
3    0  
4    0  

[5 rows x 335 columns]

In [153]:
toronto_onehot.shape

(4878, 335)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [154]:
toronto_grouped = toronto_onehot.groupby('Neighbourhood').mean().reset_index()
toronto_grouped

Neighbourhood  Accessories Store  \
0                                         Agincourt                0.0   
1                            Alderwood, Long Branch                0.0   
2   Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                   Bayview Village                0.0   
4                 Bedford Park, Lawrence Manor East                0.0   
..                                              ...                ...   
93                      Willowdale, Willowdale West                0.0   
94                                           Woburn                0.0   
95                                 Woodbine Heights                0.0   
96                                  York Mills West                0.0   
97                         York Mills, Silver Hills                0.0   

    Afghan Restaurant  African Restaurant  Airport  American Restaurant  \
0                 0.0                 0.0      0.0              0.00000   
1                 0.0                 0.0      0.0              0.00000   
2                 0.0                 0.0      0.0              0.00000   
3                 0.0                 0.0      0.0              0.00000   
4                 0.0                 0.0      0.0              0.02439   
..                ...                 ...      ...                  ...   
93                0.0                 0.0      0.0              0.00000   
94                0.0                 0.0      0.0              0.00000   
95                0.0                 0.0      0.0              0.00000   
96                0.0                 0.0      0.0              0.00000   
97                0.0                 0.0      0.0              0.00000   

    Amphitheater  Animal Shelter  Antique Shop  Aquarium  ...  Video Store  \
0            0.0             0.0           0.0       0.0  ...      0.00000   
1            0.0             0.0           0.0       0.0  ...      0.00000   
2            0.0             0.0           0.0       0.0  ...      0.00000   
3            0.0             0.0           0.0       0.0  ...      0.00000   
4            0.0             0.0           0.0       0.0  ...      0.02439   
..           ...             ...           ...       ...  ...          ...   
93           0.0             0.0           0.0       0.0  ...      0.00000   
94           0.0             0.0           0.0       0.0  ...      0.00000   
95           0.0             0.0           0.0       0.0  ...      0.00000   
96           0.0             0.0           0.0       0.0  ...      0.00000   
97           0.0             0.0           0.0       0.0  ...      0.00000   

    Vietnamese Restaurant  Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  \
0                    0.02              0.0         0.0       0.0        0.0   
1                    0.00              0.0         0.0       0.0        0.0   
2                    0.00              0.0         0.0       0.0        0.0   
3                    0.00              0.0         0.0       0.0        0.0   
4                    0.00              0.0         0.0       0.0        0.0   
..                    ...              ...         ...       ...        ...   
93                   0.00              0.0         0.0       0.0        0.0   
94                   0.00              0.0         0.0       0.0        0.0   
95                   0.00              0.0         0.0       0.0        0.0   
96                   0.00              0.0         0.0       0.0        0.0   
97                   0.00              0.0         0.0       0.0        0.0   

    Wings Joint  Women's Store  Yoga Studio  Zoo  
0       0.00000            0.0          0.0  0.0  
1       0.00000            0.0          0.0  0.0  
2       0.00000            0.0          0.0  0.0  
3       0.00000            0.0          0.0  0.0  
4       0.02439            0.0          0.0  0.0  
..          ...            ...          ...  ...  
93   

In [155]:
toronto_grouped.shape

(98, 335)

Getting the top 10 for each neighbourhood

In [156]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [157]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt    Chinese Restaurant   
1                           Alderwood, Long Branch        Discount Store   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village                  Bank   
4                Bedford Park, Lawrence Manor East    Italian Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0         Shopping Mall                Bakery           Coffee Shop   
1              Pharmacy     Convenience Store           Pizza Place   
2                  Bank    Frozen Yogurt Shop               Dog Run   
3         Grocery Store   Japanese Restaurant           Gas Station   
4           Coffee Shop                  Park           Pizza Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Caribbean Restaurant        Sandwich Place                Lounge   
1           Gas Station         Shopping Mall          Liquor Store   
2           Gas Station    Chinese Restaurant      Sushi Restaurant   
3    Chinese Restaurant                  Park            Restaurant   
4        Sandwich Place                  Bank                  Café   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0          Skating Rink  Latin American Restaurant       Sushi Restaurant  
1            Donut Shop                       Park         Sandwich Place  
2           Supermarket  Middle Eastern Restaurant                  Trail  
3                  Café               Skating Rink                  Trail  
4            Bagel Shop            Thai Restaurant                 Bakery

#### Clustering k=5

In [158]:
# import k-means from clustering stage
from sklearn.cluster import KMeans

# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:100]

array([3, 3, 3, 3, 2, 2, 0, 2, 2, 2, 3, 2, 3, 2, 2, 2, 3, 3, 2, 2, 2, 3,
       2, 3, 3, 2, 2, 3, 2, 2, 2, 2, 3, 3, 3, 2, 2, 3, 3, 0, 3, 2, 3, 3,
       2, 3, 2, 2, 2, 2, 2, 3, 2, 2, 2, 3, 2, 1, 3, 4, 2, 3, 3, 2, 2, 2,
       0, 2, 0, 2, 3, 3, 3, 2, 2, 3, 2, 2, 2, 2, 2, 2, 2, 3, 2, 2, 3, 0,
       3, 3, 3, 2, 2, 3, 0, 2, 0, 4], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [159]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = combined_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head() # check the last columns!

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             3.0                  Park   
1  43.725882 -79.315572             3.0           Coffee Shop   
2  43.654260 -79.360636             2.0           Coffee Shop   
3  43.718518 -79.464763             2.0        Clothing Store   
4  43.662301 -79.389494             2.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Convenience Store              Bus Stop         Shopping Mall   
1           Men's Store            Boxing Gym  Gym / Fitness Center   
2                   Pub               Theater                  Park   
3  Fast Food Restaurant           Coffee Shop            Restaurant   
4                  Park               Gay Bar    Italian Restaurant   

   5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  \
0               Pharmacy     Fish & Chips Shop      Food & Drink Shop   
1  Portuguese Restaurant   Sporting Goods Shop            Golf Course   
2                   Café        Breakfast Spot                 Bakery   
3    Fried Chicken Joint          Dessert Shop  Vietnamese Restaurant   
4       Sushi Restaurant           Pizza Place                   Café   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Café  Fast Food Restaurant            Supermarket  
1            Hockey Arena          Intersection             Playground  
2                   Diner            Restaurant  Performing Arts Venue  
3  Furniture / Home Store      Sushi Restaurant      Accessories Store  
4        Ramen Restaurant             Gastropub            Yoga Studio

Check for NaN values and drop them to prevent data skew

In [160]:
toronto_merged[toronto_merged['Cluster Labels'].isnull()]

Postal Code      Borough Neighbourhood   Latitude  Longitude  \
95         M1X  Scarborough   Upper Rouge  43.836125 -79.205636   

    Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
95             NaN                   NaN                   NaN   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
95                   NaN                   NaN                   NaN   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
95                   NaN                   NaN                   NaN   

   9th Most Common Venue 10th Most Common Venue  
95                   NaN                    NaN

In [161]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Plot the clusters in the map

In [162]:
import matplotlib.cm as cm
import matplotlib.colors as colors

In [163]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Examine Clusters

#### Cluster 1

Cluster 1 appears to have alot of parks and playgrounds. We will call this one 'Recreation'.

In [164]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
11         Etobicoke             0.0                  Park   
12       Scarborough             0.0            Playground   
22       Scarborough             0.0           Coffee Shop   
57        North York             0.0         Auto Workshop   
58       Scarborough             0.0                  Park   
66        North York             0.0                  Park   
91  Downtown Toronto             0.0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
11           Pizza Place            Restaurant                  Bank   
12          Burger Joint    Italian Restaurant                  Park   
22                  Park    Chinese Restaurant  Fast Food Restaurant   
57                  Park           Golf Course     Convenience Store   
58     Convenience Store         Auto Workshop              Gym Pool   
66            Restaurant           Coffee Shop                   Gym   
91                  Park         Grocery Store        Breakfast Spot   

   5th Most Common Venue  6th Most Common Venue 7th Most Common Venue  \
11         Grocery Store                    Gym    Mexican Restaurant   
12        Breakfast Spot      Food & Drink Shop                 Field   
22     Indian Restaurant      Mobile Phone Shop           Escape Room   
57                Bakery           Intersection      Storage Facility   
58                   Gym  General Entertainment                 Diner   
66             Pet Store                Dog Run    Chinese Restaurant   
91                Bistro                   Bank        Sandwich Place   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
11        Clothing Store     Fish & Chips Shop                  Hotel  
12           Escape Room  Ethiopian Restaurant            Event Space  
22  Ethiopian Restaurant           Event Space     Falafel Restaurant  
57        Discount Store           Gas Station                    Zoo  
58            Restaurant       College Stadium           Skating Rink  
66         Bowling Alley            Playground          Grocery Store  
91             BBQ Joint   Filipino Restaurant     Athletics & Sports

#### Cluster 2

Cluster 2 only contains one neighborhood, and it has a Zoo and Lounges in it. We are going to call it 'Zoo/Entertainment'

In [165]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
94  Etobicoke             1.0                Lounge           Coffee Shop   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
94                   Zoo                 Field  Ethiopian Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
94           Event Space    Falafel Restaurant                  Farm   

   9th Most Common Venue 10th Most Common Venue  
94        Farmers Market   Fast Food Restaurant

#### Cluster 3

Cluster 3 is a large cluster with many coffee shops and restraunts. We will call this one 'Dining'.

In [166]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2    Downtown Toronto             2.0           Coffee Shop   
3          North York             2.0        Clothing Store   
4    Downtown Toronto             2.0           Coffee Shop   
6         Scarborough             2.0           Coffee Shop   
7          North York             2.0            Restaurant   
9    Downtown Toronto             2.0           Coffee Shop   
13         North York             2.0            Restaurant   
14          East York             2.0           Coffee Shop   
15   Downtown Toronto             2.0                  Café   
19       East Toronto             2.0                   Pub   
20   Downtown Toronto             2.0           Coffee Shop   
23          East York             2.0   Sporting Goods Shop   
24   Downtown Toronto             2.0           Coffee Shop   
25   Downtown Toronto             2.0     Korean Restaurant   
30   Downtown Toronto             2.0           Coffee Shop   
31       West Toronto             2.0                  Café   
33         North York             2.0           Coffee Shop   
35          East York             2.0                  Café   
36   Downtown Toronto             2.0           Coffee Shop   
37       West Toronto             2.0                  Café   
41       East Toronto             2.0      Greek Restaurant   
42   Downtown Toronto             2.0           Coffee Shop   
43       West Toronto             2.0           Coffee Shop   
47       East Toronto             2.0     Indian Restaurant   
48   Downtown Toronto             2.0           Coffee Shop   
52         North York             2.0     Korean Restaurant   
54       East Toronto             2.0           Coffee Shop   
55         North York             2.0    Italian Restaurant   
59         North York             2.0           Coffee Shop   
61    Central Toronto             2.0                  Park   
62    Central Toronto             2.0    Italian Restaurant   
67    Central Toronto             2.0           Coffee Shop   
68    Central Toronto             2.0                  Park   
69       West Toronto             2.0                  Café   
73    Central Toronto             2.0    Italian Restaurant   
74    Central Toronto             2.0    Italian Restaurant   
75       West Toronto             2.0                  Café   
76        Mississauga             2.0           Coffee Shop   
79    Central Toronto             2.0           Coffee Shop   
80   Downtown Toronto             2.0                  Café   
81       West Toronto             2.0           Coffee Shop   
83    Central Toronto             2.0           Coffee Shop   
84   Downtown Toronto             2.0                   Bar   
86    Central Toronto             2.0           Coffee Shop   
87   Downtown Toronto             2.0           Coffee Shop   
88          Etobicoke             2.0                  Park   
92   Downtown Toronto             2.0           Coffee Shop   
96   Downtown Toronto             2.0                  Café   
97   Downtown Toronto             2.0           Coffee Shop   
98          Etobicoke             2.0           Coffee Shop   
99   Downtown Toronto             2.0           Coffee Shop   
100      East Toronto             2.0                  Park   
102         Etobicoke             2.0            Restaurant   

    2nd Most Common Venue          3rd Most Common Venue  \
2                     Pub                        Theater   
3    Fast Food Restaurant                    Coffee Shop   
4                    Park                        Gay Bar   
6                   Trail           Fast Food Restaurant   
7             Coffee Shop            Japanese Restaurant   
9               Gastropub             Italian Restaurant   
13            Coffee Shop            Japanese Restaurant   
14                   Park                           Café   
15            Coffee Shop            Japanese Restaurant   
19         

#### Cluster 4

Cluster 4 looks like it has a lot of shopping and dining in it. Probably Mostly residential based on the tupes of places we are seeing. We will go with 'Residential'.

In [167]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels      1st Most Common Venue  \
0    North York             3.0                       Park   
1    North York             3.0                Coffee Shop   
5     Etobicoke             3.0                   Pharmacy   
8     East York             3.0       Gym / Fitness Center   
10   North York             3.0              Grocery Store   
16         York             3.0                Pizza Place   
17    Etobicoke             3.0                Coffee Shop   
18  Scarborough             3.0                Pizza Place   
21         York             3.0                   Pharmacy   
26  Scarborough             3.0                Coffee Shop   
27   North York             3.0                   Pharmacy   
28   North York             3.0                Coffee Shop   
29    East York             3.0                Coffee Shop   
32  Scarborough             3.0             Ice Cream Shop   
34   North York             3.0                Coffee Shop   
38  Scarborough             3.0                Coffee Shop   
39   North York             3.0                       Bank   
40   North York             3.0                Coffee Shop   
44  Scarborough             3.0                   Bus Line   
46   North York             3.0                Coffee Shop   
49   North York             3.0                Coffee Shop   
50   North York             3.0          Electronics Store   
51  Scarborough             3.0                Pizza Place   
53   North York             3.0                Coffee Shop   
56         York             3.0     Furniture / Home Store   
60   North York             3.0                Coffee Shop   
63         York             3.0                Coffee Shop   
64         York             3.0                Pizza Place   
65  Scarborough             3.0          Electronics Store   
70    Etobicoke             3.0                Pizza Place   
71  Scarborough             3.0  Middle Eastern Restaurant   
72   North York             3.0                   Pharmacy   
77    Etobicoke             3.0                   Pharmacy   
78  Scarborough             3.0         Chinese Restaurant   
82  Scarborough             3.0                Coffee Shop   
85  Scarborough             3.0         Chinese Restaurant   
89    Etobicoke             3.0                Pizza Place   
90  Scarborough             3.0         Chinese Restaurant   
93    Etobicoke             3.0             Discount Store   

     2nd Most Common Venue       3rd Most Common Venue  \
0        Convenience Store                    Bus Stop   
1              Men's Store                  Boxing Gym   
5                   Bakery                        Bank   
8              Pizza Place  Construction & Landscaping   
10    Fast Food Restaurant                 Gas Station   
16             Coffee Shop           Convenience Store   
17          Farmers Market                    Pharmacy   
18    Fast Food Restaurant                        Bank   
21                    Park                 Pizza Place   
26                Pharmacy                        Bank   
27             Coffee Shop                        Park   
28                    Bank          Frozen Yogurt Shop   
29       Indian Restaurant               Grocery Store   
32           Grocery Store                  Restaurant   
34  Furniture / Home Store                 Pizza Place   
38      Chinese Restaurant               Grocery Store   
39           Grocery Store         Japanese Restaurant   
40   Vietnamese Restaurant                       Hotel   
44             Coffee Shop                      Bakery   
46   Vietnamese Restaurant                       Hotel   
49                  Bakery                        Park   
50      Italian Restaurant                        Bank   
51          Ice Cream Shop                       Beach   
53   Vietnamese Restaurant                       Hotel   
56           Grocery Store          Italian Restaurant   
60   Vietnamese Res

#### Cluster 5

Cluster 5 is an odd ball. We have Parks again, but also a pool and Elementary Schools. We will call it 'Other' for lack of a better definition.

In [168]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45   North York             4.0                  Park                  Pool   
101   Etobicoke             4.0                  Park        Ice Cream Shop   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
45                    Zoo                 Field           Escape Room   
101    Italian Restaurant              Bus Stop         Shopping Mall   

           6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45          Ethiopian Restaurant           Event Space    Falafel Restaurant   
101  Eastern European Restaurant  Gym / Fitness Center          Dessert Shop   

    9th Most Common Venue 10th Most Common Venue  
45                   Farm         Farmers Market  
101         Design Studio            Event Space